core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -411,6 +427,34 @@ void connect(SocketAddress remoteAddress, SessionProtocol desiredProtocol,
         });
     }
 
+    @Nullable
+    private static SslContext maybeCreateSslContext(SocketAddress remoteAddress,
```

Comment:
Not too strong on this, but possibly it makes more sense that `tlsProvider` is provided via `ClientFactory` instead of the `WebClient` level.

The reasoning is that `TlsProvider` is invoked only on creating new connections. For this reason, calling `WebClient.tlsProvider(...).execute()` does not guarantee that the request will use the specified `tlsProvider`.

Rather, perhaps `TlsProvider` can be set on the `ClientFactory` level and `TlsKeyPair` can be set on the `Client` level. Thoughts?

Reply:
> For this reason, calling WebClient.tlsProvider(...).execute() does not guarantee that the request will use the specified tlsProvider.

Good point. Even though a `TlsProvider` is set to a `ClientFactory`, each connection may be authenticated with a different `TlsKeyPair`. Because `TlsProvider` may return different key pairs even for the same authority.

If a user wants a more sophisticated level of authentication at the request level, it would be better to use OAuth 2.0 or similar.

I agree that this can be confusing. Let's get more opinion on that.

Reply:
> Good point. Even though a TlsProvider is set to a ClientFactory, each connection may be authenticated with a different TlsKeyPair. Because TlsProvider may return different key pairs for the same authority.

Even worse, a `TlsProvider` set for `WebClientB` may seem to leak to `WebClientA`

```
val factory = ClientFactory.builder().tls(tlsKeyPairC).build()
val webClientA = WebClient.builder().tlsProvider(tlsKeyPairA).factory(factory).build();
val webClientB = WebClient.builder().tlsProvider(tlsKeyPairB).factory(factory).build();

webClientA.execute() // maybe uses `tlsKeyPairA`, `tlsKeyPairB`, `tlsKeyPairC`
```

I feel like this is contrary to the current set of APIs, which guarantees that each option for each client doesn't affect others. Hopefully this is also taken in to consideration

Reply:
`TlsProvider` is removed from `WebClient`.